src/main/java/hudson/plugins/analysis/core/HealthAwarePublisher.java

```diff
@@ -210,8 +222,33 @@ protected void updateBuildResult(final BuildResult result, final PluginLogger lo
      *             a better error message, if it can do so, so that users have
      *             better understanding on why it failed.
      */
-    protected abstract BuildResult perform(AbstractBuild<?, ?> build, PluginLogger logger)
-            throws InterruptedException, IOException;
+    @Deprecated
+    protected BuildResult perform(AbstractBuild<?, ?> build, PluginLogger logger)
+            throws InterruptedException, IOException {
+        return perform((Run) build, null, logger);
+    }
+
+    protected /*abstract*/ BuildResult perform(Run<?, ?> run, FilePath workspace, PluginLogger logger) throws InterruptedException, IOException{
+        if (isOverridden(HealthAwarePublisher.class, getClass(),
+                "perform", AbstractBuild.class, PluginLogger.class) && run instanceof AbstractBuild) {
```

Comment:
Well this code is only run in deprecated scenarios, and once per build, so a few nanoseconds here or there are not going to be missed.